tests/test.c

```diff
@@ -3113,6 +3115,106 @@ void test_consumer_verify_assignment0(const char *func,
         rd_kafka_topic_partition_list_destroy(assignment);
 }
 
+/**
+ * @brief Verify that the consumer's assignment matches the expected assignment.
+ *        passed as a topic partition list in \p expected_assignment .
+ */
+rd_bool_t test_consumer_verify_assignment_topic_partition_list0(
+    const char *func,
+    int line,
+    rd_kafka_t *rk,
+    const rd_kafka_topic_partition_list_t *expected_assignment) {
+        rd_kafka_topic_partition_list_t *assignment, *expected_assignment_copy;
+        rd_kafka_resp_err_t err;
+        int i;
+        rd_bool_t ret = rd_true;
+
+        if ((err = rd_kafka_assignment(rk, &assignment)))
+                TEST_FAIL("%s:%d: Failed to get assignment for %s: %s", func,
+                          line, rd_kafka_name(rk), rd_kafka_err2str(err));
+
+        TEST_SAYL(4, "%s assignment (%d partition(s)):\n", rd_kafka_name(rk),
+                  assignment->cnt);
+        for (i = 0; i < assignment->cnt; i++)
+                TEST_SAYL(4, " %s [%" PRId32 "]\n", assignment->elems[i].topic,
+                          assignment->elems[i].partition);
+
+        if (assignment->cnt != expected_assignment->cnt) {
+                ret = rd_false;
+                goto done;
+        }
+
+        expected_assignment_copy =
+            rd_kafka_topic_partition_list_copy(expected_assignment);
+        rd_kafka_topic_partition_list_sort(assignment, NULL, NULL);
+        rd_kafka_topic_partition_list_sort(expected_assignment_copy, NULL,
+                                           NULL);
+
+        for (i = 0; i < assignment->cnt; i++) {
+                if (strcmp(assignment->elems[i].topic,
+                           expected_assignment_copy->elems[i].topic) ||
+                    assignment->elems[i].partition !=
+                        expected_assignment_copy->elems[i].partition) {
+                        ret = rd_false;
+                        goto done;
+                }
+        }
+
+done:
+        rd_kafka_topic_partition_list_destroy(assignment);
+        return ret;
+}
+
+/**
+ * @brief Wait until the consumer's assignment matches the expected assignment.
+ *        passed as a topic partition list in \p expected_assignment .
+ *        Polling if \p do_poll is true, otherwise sleeps.
+ *        Until \p timeout_ms milliseconds.
+ */
+void test_consumer_wait_assignment_topic_partition_list0(
+    const char *func,
+    int line,
+    rd_kafka_t *rk,
+    rd_bool_t do_poll,
+    const rd_kafka_topic_partition_list_t *expected_assignment,
+    int timeout_ms) {
+        int i;
+        rd_ts_t end        = test_clock() + timeout_ms * 1000;
+        rd_bool_t verified = rd_false;
+        rd_kafka_topic_partition_list_t *expected_assignment_copy;
+
+        TEST_SAY("Verifying assignment\n");
+        expected_assignment_copy =
+            rd_kafka_topic_partition_list_copy(expected_assignment);
```

Comment:
We don't need this copy as well. It is already being copied in `test_consumer_verify_assignment_topic_partition_list0`. It is expected that the called function doesn't modify the list.